### PR TITLE
Adhere to the consuming builder pattern for SagaTemplateBuilder

### PR DESCRIPTION
**Pull request by smklein**
For context, see: https://doc.rust-lang.org/1.0.0/style/ownership/builders.html

TL;DR: there are two patterns of builders that are common in Rust.

# Non-Consuming

- `append` (and `append_parallel`) would act on a `&mut self` and return `&mut self`. This allows chaining.
- `build` would act on a `&mut self` and return a `SagaTemplate<UserType>`. This allows chaining, but does not consume the builder on creation.

# Consuming

- `append` (and `append_parallel`) would act on a `mut self` and return `mut self`. This allows chaining.
- `build` would act on a `mut self` and return a `SagaTemplate<UserType>`. This allows chaining while consuming the builder on creation.

This PR updates the `SagaTemplateBuilder` to be a "chainable, consuming builder" - I chose "consuming" based on the current signature to `build`, which consumes `mut self`, though the non-consuming variant would also be possible if this was relaxed to `&mut self`.

### CODE REVIEW

**Comment by davepacheco** [comment]
I did not read that document to imply that if the builder's methods consume `self`, then `build()` really ought to consume `self` and vice versa.  Is it important to constrain builders to those two modes?

The thing I don't love about the consuming pattern is that you _have_ to chain calls.  I've found this particularly awkward when you want to use one the builder methods conditionally because I wind up doing:

```
let builder = Builder::new()
    .first_thing(...);

let builder = if (condition) {
    builder.conditionally_do_second_thing()
} else {
    builder
}

...
```

Is there a better pattern for that?

**Comment by smklein** [comment]
> I did not read that document to imply that if the builder's methods consume `self`, then `build()` really ought to consume `self` and vice versa. Is it important to constrain builders to those two modes?

I thought this was implied in the document by the following (within the "Consuming builders" section):

"Sometimes builders must transfer ownership when constructing the final type T, meaning that the terminal methods must take self rather than &self... Under the rubric of making easy things easy and hard things possible, all builder methods for a consuming builder should take and returned an owned self."

I interpreted this as:
- Given the constraint of a consuming `build`...
- ... The builder methods should act on an owned `self`.

> The thing I don't love about the consuming pattern is that you _have_ to chain calls. I've found this particularly awkward when you want to use one the builder methods conditionally because I wind up doing:
> 
> ```
> let builder = Builder::new()
>     .first_thing(...);
> 
> let builder = if (condition) {
>     builder.conditionally_do_second_thing()
> } else {
>     builder
> }
> 
> ...
> ```
> 
> Is there a better pattern for that?

I think this would be possible succinctly with the following:

```rust
let mut builder = Builder::new();

if condition {
  builder = builder.conditionally_do_second_thing()
}
// No need to re-assign "builder" if it isn't modified
```

**Comment by davepacheco** [comment]
> > The thing I don't love about the consuming pattern is that you _have_ to chain calls. I've found this particularly awkward when you want to use one the builder methods conditionally because I wind up doing:
> > ```
> > let builder = Builder::new()
> >     .first_thing(...);
> > 
> > let builder = if (condition) {
> >     builder.conditionally_do_second_thing()
> > } else {
> >     builder
> > }
> > 
> > ...
> > ```
> > 
> > 
> > Is there a better pattern for that?
> 
> I think this would be possible succinctly with the following:
> 
> ```rust
> let mut builder = Builder::new();
> 
> if condition {
>   builder = builder.conditionally_do_second_thing()
> }
> // No need to re-assign "builder" if it isn't modified
> ```
> 
> Due to the fact that the builder methods don't change the type of the `builder` object, so the re-assignment isn't necessary in the no-op case.

That is fair, but also seems more awkward than:

```
if condition {
    builder.conditionally_do_second_thing()
}
```

**Comment by smklein** [comment]
> > > The thing I don't love about the consuming pattern is that you _have_ to chain calls. I've found this particularly awkward when you want to use one the builder methods conditionally because I wind up doing:
> > > ```
> > > let builder = Builder::new()
> > >     .first_thing(...);
> > > 
> > > let builder = if (condition) {
> > >     builder.conditionally_do_second_thing()
> > > } else {
> > >     builder
> > > }
> > > 
> > > ...
> > > ```
> > > 
> > > 
> > > Is there a better pattern for that?
> > 
> > 
> > I think this would be possible succinctly with the following:
> > ```rust
> > let mut builder = Builder::new();
> > 
> > if condition {
> >   builder = builder.conditionally_do_second_thing()
> > }
> > // No need to re-assign "builder" if it isn't modified
> > ```
> > 
> > 
> > Due to the fact that the builder methods don't change the type of the `builder` object, so the re-assignment isn't necessary in the no-op case.
> 
> That is fair, but also seems more awkward than:
> 
> ```
> if condition {
>     builder.conditionally_do_second_thing()
> }
> ```

Agreed! I think this is why "non-consuming" builders are the preferred pattern - they act upon `&mut self`, and don't require the reassignment.

Admittedly, we could split the difference here and have "non-consuming argument methods with a consuming builder pattern", like the following:

```rust
pub fn new() -> Self { ... }
pub fn append(&mut self, ...) -> &mut self { ... }
pub fn append_parallel(&mut self, ...) -> &mut self { ... }
pub fn build(mut self) -> ...
```

This would *sort of* allow chaining - it would let the "append" methods be chained together, but would require a separate invocation for `new` and `build`, like the following:

```rust
let mut builder = SagaTemplateBuilder::new();
builder.append(...)
   .append(...)
   .append_parallel(...);
builder.build()
```

Would you prefer this option? The advantage of the fully consistent "all-consuming" or "all non-consuming" pattern is the production of templates like the following:

```rust
let output = SagaTemplateBuilder::new()
   .append(...)
   .append(...)
   .append_parallel(...);
   .build();
```

**Comment by davepacheco** [comment]
> Agreed! I think this is why "non-consuming" builders are the preferred pattern - they act upon `&mut self`, and don't require the reassignment.
> 
> Admittedly, we could split the difference here and have "non-consuming argument methods with a consuming builder pattern", like the following:
> 
> ```rust
> pub fn new() -> Self { ... }
> pub fn append(&mut self, ...) -> &mut self { ... }
> pub fn append_parallel(&mut self, ...) -> &mut self { ... }
> pub fn build(mut self) -> ...
> ```
> 
> This would _sort of_ allow chaining - it would let the "append" methods be chained together, but would require a separate invocation for `new` and `build`, like the following:
> 
> ```rust
> let mut builder = SagaTemplateBuilder::new();
> builder.append(...)
>    .append(...)
>    .append_parallel(...);
> builder.build()
> ```
> 
> Would you prefer this option? The advantage of the fully consistent "all-consuming" or "all non-consuming" pattern is the production of templates like the following:
> 
> ```rust
> let output = SagaTemplateBuilder::new()
>    .append(...)
>    .append(...)
>    .append_parallel(...)
>    .build();
> ```

Agreed that having `append` return `&mut self` would be better than returning nothing here, since it seems strictly more useful.  And also that it's a little awkward because you can't build in a one-liner.

So we've got:

* non-consuming: no technical issues, but feels conceptually misleading to me because in my experience, the builder is almost always created once for each thing that it builds, so I kind of expect it to look like the typestate pattern where once you're done building, the builder is gone.
* consuming: _have_ to chain, even when it's more convenient not to
* "split the difference" as you suggested (have `append()` return `&mut self`): the awkwardness of being able to chain most of the time, but not with `new`

Honestly, I don't think I care between these options.  They all seem slightly annoying and also fine.

**Comment by davepacheco** [comment]
Also useful context: I expect the `SagaTemplateBuilder` will probably change substantially when we get more real-world sagas prototyped.  For example, consumers may want a sequence of actions to run in parallel with some other action.  There's no way to do that today.  There are many other ways to build such an interface, with varying levels of boilerplate and flexibility.  I think the real-world uses will drive what makes sense here.

**Comment by smklein** [comment]
Also, as a citation in the stdlib for these patterns:

- [std::process::Command](https://doc.rust-lang.org/std/process/struct.Command.html) follows the non-consuming builder pattern
- [std::thread::Builder](https://doc.rust-lang.org/std/thread/struct.Builder.html) follows the consuming builder pattern.
- I can't find any examples in the stdlib of any that "split the difference".

**Comment by sunshowers** [comment]
Quick summary of a longer discussion in #oxide-rust -- I think a builder with the individual methods being `&mut self` and a final `build(&self)` method is fine, and it's the most common way I write builders in my code. I think consuming builders aren't great because they can't easily be nested inside other structs [1], which (to me) is a substantive argument that wins over what I think are some of the more trivial annoyances with nonconsuming builders.

[1] (You can nest them, but you have to either use an `Option` or `Default` and take the builder before putting it back in. This means a new check that's outside the type system.)

**Comment by davepacheco** [comment]
Thanks @sunshowers.  I'm going to close this anyway because the code's changed so much in the last three years but I'm happy to re-evaluate if we feel it's worthwhile!
